### PR TITLE
INSP: Support child modules in unused import inspection

### PR DESCRIPTION
**Pull request by dima74**
TODO: Support text search inside macro expansions

changelog: Support finding usages in child modules in unused import inspection

### CODE REVIEW

File: src/test/kotlin/org/rust/ide/inspections/lints/RsUnusedImportInspectionTest.kt [hunk 1]
```diff
@@ -662,20 +662,170 @@ class RsUnusedImportInspectionTest : RsInspectionsTestBase(RsUnusedImportInspect
...
+
+    @MockEdition(CargoWorkspace.Edition.EDITION_2018)
+    fun `test private import used in child module (glob import)`() = checkByText("""
+        mod foo {
+            pub struct S;
+        }
+        use foo::S;
+
+        mod bar {
+            use super::S;
```
Comment (Kobzol) [inline src/test/kotlin/org/rust/ide/inspections/lints/RsUnusedImportInspectionTest.kt:708]: This test is the same as `test private import used in child module (named import)`. Probably this should be `use super::*`.

**Comment by Kobzol** [comment]
After https://github.com/intellij-rust/intellij-rust/pull/7495 gets merged, use specks can be also checked for "unusedness" from outside code, not only from the `RsUnusedImportInspection`. Before calling `RsUseSpeck.isUsed`, `RsUnusedImportInspection::isApplicableForUseItem` should be called on its parent use item, otherwise the result might not be correct - especially this line: https://github.com/intellij-rust/intellij-rust/pull/7550/files#diff-2b3d285d045313a514e8b014bb459baa9bff1298322cc65db2575cb5e4a1d468R128 should probably return `true` and not raise an exception to be sure.

However, the separation of `isUsed` and `isApplicableForUseItem` was done for (probably not even needed) performance reasons. But actually with this PR the `isUsed` function could  probably just do the check always when called, since it will just check if the use item is public or not (and it no longer has to check the presence of child modules).

**Comment by Kobzol** [comment]
Btw, the `regressions` workflow probably doesn't run this inspection, right? Maybe we should enable it to find possible regressions (especially after this PR, which enables the inspection in more cases).

**Comment by dima74** [comment]
> However, the separation of `isUsed` and `isApplicableForUseItem` was done for (probably not even needed) performance reasons. But actually with this PR the `isUsed` function could probably just do the check always when called, since it will just check if the use item is public or not (and it no longer has to check the presence of child modules).

Thanks, I added `isApplicableForUseItem` check to `isUsed`

**Comment by dima74** [comment]
Causes of regressions:
* tokio - one file belongs to different crates (via `include!` macro or module with `#[path]` attribute)
* amethyst - for some reason proc macros are not expanded (there are no regressions in real IDE if proc macros are enabled)

File: src/main/kotlin/org/rust/ide/inspections/lints/RsUnusedImportInspection.kt [hunk 2]
```diff
@@ -132,13 +121,22 @@ private fun getHighlightElement(useSpeck: RsUseSpeck): PsiElement {
...
+}
+
 /**
  * Returns true if this use speck has at least one usage in its containing owner (module or function).
  * A usage can be either a path that uses the import of the use speck or a method call/associated item available through
  * a trait that is imported by this use speck.
  */
 fun RsUseSpeck.isUsed(): Boolean {
-    val owner = this.parentOfType<RsUseItem>()?.parent as? RsItemsOwner ?: return true
+    val useItem = this.parentOfType<RsUseItem>() ?: return true
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/lints/RsUnusedImportInspection.kt:137]:
```suggestion
    val useItem = ancestorStrict<RsUseItem>() ?: return true
```

File: src/main/kotlin/org/rust/ide/inspections/lints/RsUnusedImportInspection.kt [hunk 3]
```diff
@@ -156,11 +154,113 @@ private fun isUseSpeckUsed(useSpeck: RsUseSpeck, usage: PathUsageMap): Boolean {
...
+private fun isItemUsedInSameMod(item: RsNamedElement, importName: String, usage: PathUsageMap): Boolean {
+    val used = item in usage.pathUsages[importName].orEmpty()
+        || item in usage.traitUsages
+    val probablyUsed = importName in usage.unresolvedPaths
+        || (item as? RsTraitItem)?.expandedMembers.orEmpty().any { it.name in usage.unresolvedMethods }
+    return used || probablyUsed
+}
+
+private fun isItemUsedInOtherMods(item: RsNamedElement, importName: String, useSpeck: RsUseSpeck): Boolean {
+    val useItem = useSpeck.parentOfType<RsUseItem>() ?: return true
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/lints/RsUnusedImportInspection.kt:171]:
```suggestion
    val useItem = useSpeck.ancestorStrict<RsUseItem>() ?: return true
```

File: src/main/kotlin/org/rust/ide/inspections/lints/RsUnusedImportInspection.kt [hunk 4]
```diff
@@ -156,11 +154,113 @@ private fun isUseSpeckUsed(useSpeck: RsUseSpeck, usage: PathUsageMap): Boolean {
...
+    searchScope: SearchScope?,
+    identifier: String,
+    originalProcessor: (RsReferenceElement) -> Boolean
+): Boolean {
+    // returning `false` stops the processing
+    fun processor(element: PsiElement): Boolean {
+        if (element !is RsReferenceElementBase || element.referenceName != identifier) return true
+        element.findExpansionElements()?.let { expansionElements ->
+            return expansionElements
+                .mapNotNull { it.parentOfType<RsElement>() }  // PsiElement(identifier)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/lints/RsUnusedImportInspection.kt:250]:
```suggestion
                .mapNotNull { it.ancestorStrict<RsElement>() }  // PsiElement(identifier)
```

**Review by vlad20012 — COMMENTED** [review]
Let's add a test where this doesn't work with macros

**Review by vlad20012 — APPROVED** [review]
:+1: 

bors r+

**Pull request merged**